### ISSUE TITLE
MPI error on CURTA cluster with 2DCartesian geometry

### ISSUE THREAD

**Issue by lesnat** [SmileiPIC/Smilei#270]
## Description

Hi everyone !
Hope everything is fine for you.

I have some troubles with a 2D simulation on the CURTA cluster: when I launch the simulation I get the following error 

```
 Applying prescribed fields at time t = 0
 --------------------------------------------------------------------------------
Abort(940160260) on node 7 (rank 7 in comm 0): Fatal error in PMPI_Irecv: Invalid tag, error stack:
PMPI_Irecv(156): MPI_Irecv(buf=0x56763840, count=150, MPI_DOUBLE, src=1, tag=2751111, MPI_COMM_WORLD, request=0x4e0d450) failed
PMPI_Irecv(100): Invalid tag, value is 2751111
Abort(269071620) on node 8 (rank 8 in comm 0): Fatal error in PMPI_Irecv: Invalid tag, error stack:
PMPI_Irecv(156): MPI_Irecv(buf=0x328d01c0, count=150, MPI_DOUBLE, src=1, tag=2341511, MPI_COMM_WORLD, request=0x28563700) failed
PMPI_Irecv(100): Invalid tag, value is 2341511
Abort(738833668) on node 2 (rank 2 in comm 0): Fatal error in PMPI_Irecv: Invalid tag, error stack:
PMPI_Irecv(156): MPI_Irecv(buf=0x315d090, count=150, MPI_DOUBLE, src=1, tag=4894611, MPI_COMM_WORLD, request=0x946fc270) failed
PMPI_Irecv(100): Invalid tag, value is 4894611
Abort(636164) on node 1 (rank 1 in comm 0): Fatal error in PMPI_Irecv: Invalid tag, error stack:
PMPI_Irecv(156): MPI_Irecv(buf=0x27e89e0, count=150, MPI_DOUBLE, src=0, tag=4894411, MPI_COMM_WORLD, request=0x170817dc0) failed
PMPI_Irecv(100): Invalid tag, value is 4894411
```
## Steps to reproduce the problem

Here is the namelist I'm using.

[namelist.py.txt](https://github.com/SmileiPIC/Smilei/files/4851817/namelist.py.txt)

I used the `smilei_test` app and everything seems to works well, but the error occurs at field initialization with Smilei version 4.4 (v-854724ce) or 4.5 (v-3a7ca3b4), compiled without vectorization or any advanced option. I also tried to comment the diagnostics on the namelist but I get the same error. I made 1D simulations on CURTA some times ago (with the same Smilei apps as I remember) but I had no problems to run it.

I'm using Intel MPI version 2019 update 3 build 20190214 (id: b645a4a54), with HDF5 v-1.10.5, Python 2.7.5 and g++ v-4.8.5 20150623 (Red Hat 4.8.5-36).

I also tried the same namelist on my local computer with smilei v4.1 and it seems to run with no problem.

The problem seems to come from MPI but I have no idea of why it seems not work only on 2D geometry. I can make more tests if you need it !

Thank's a lot for your help,
Cheers,
Léo.

**Comment by jderouillat** [comment]
Hi Léo,
How many MPI processus are you using ?  And how many OpenMP threads ?

**Comment by jderouillat** [comment]
I reproduced it on a recent computer, not an old.  
I'll inform you as soon as possible.

**Comment by jderouillat** [comment]
It seems that in Intel-mpi 2019-3, there is too many patches 512 x 256 for too few MPI processes, below the `patch_load.txt` (number of patches per process) :
```bash
patch count = 48947
patch count = 48947
patch count = 6991
patch count = 3263
patch count = 3263
patch count = 3263
patch count = 3263
patch count = 4936
patch count = 4936
patch count = 3263
```
With 128 x 64 (maximum number of patches on a process : 3721), it runs.  
With IntelMPI 2019, the tag max seems to be 1048575 while with an older IntelMPI I find 536870911 using `MPI_Comm_get_attr` like in #62 .

Tags are computed concatenating the local patch id and some other digits (1 for the direction, 1 for sense in the direction, 1 additional for some kind of data). This is to explain why you have these tags : 48944-1-1.

This is not satisfying regarding Intel 2019 but at least it explains.  
Regarding the code, so many patches per process will produce large overhead in sync time.

I'm looking to a solution for IntelPI 2019.

**Comment by jderouillat** [comment]
With the Update 7 Build 20200312 of IntelMPI (available on Jean Zay at IDRIS), the max tag value is much more larger : 1073741823.

**Comment by lesnat** [comment]
Thank's a lot for your time !

I tried with Intel MPI 2019.5.281 (also available on CURTA) with 4 MPI processes and it seems to work ! (at least it initializes the fields and run few hundred timesteps).

The version 2018.5.274 of Intel MPI is also available at CURTA. Want you that I try with this MPI version ?

Anyway thank's again, I'll use the 2019.5.281 version of Intel MPI and let you know if I have other troubles :)

Have a nice day !
Léo.

**Comment by jderouillat** [comment]
We will not look longer for a bug fix for these kind of MPI libraries which are very restrictive.  
But we propose to introduce a test on the max value of a tag of the library and of the max tag value generated by the code. Regarding those values, the code will suggest explicitly (or not) to use more MPI process, less patches or to change of MPI library.  
Are you ok with this ?

**Comment by lesnat** [comment]
Yes it's ok for me ! Thank's :)

**Comment by jderouillat** [comment]
The proposed test has been implemented in c11cedfef437cff0b6ad17039f026b3d25261522.
